### PR TITLE
adding average point separation postprocessor

### PR DESCRIPTION
**Pull request by jain651**
# Reason
Currently to calculate deformation between two points in a domain, we need to get displacement in each direction for individual points and then manually obtain the deformation.

# Design
A new postprocessor is required to obtain deformation between two points.

# Impact
Reduces the length of input file.
#18751

### CODE REVIEW

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/src/postprocessors/AveragePointSeparation.C [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+AveragePointSeparation::execute()
+{
+  _value = 0.;
+  Real sq_diff;
+  for (unsigned int i = 0; i < _first_point.size(); ++i)
+  {
+    sq_diff = 0;
+    for (unsigned int j = 0; j < _displacements.size(); ++j)
+      sq_diff += Utility::pow<2>(_system.point_value(_disp_num[j], _first_point[i], false) -
+                                 _system.point_value(_disp_num[j], _last_point[i], false));
```
[the remark this answers is not in the document]
Comment (jain651) [reply modules/tensor_mechanics/src/postprocessors/AveragePointSeparation.C:77]: I agree with the first point. I was trying to get the average deformation, but you are right I should use the total separation in the calculation. 

For second and third points:
Here, I am trying to do what we generally do in experiment. We measure the separation between a set of first points and last points, then take take an average of measurements to obtain average deformation. I am trying to do the same thing. 
For this reason, both array should have the same size. I will put a check for this too in the corrected PR.

Please let me know if you still think that we should take the average of all first point and last point array and then get the average separation.

File: modules/tensor_mechanics/doc/content/source/postprocessors/AveragePointSeparation.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# AveragePointSeparation
+
+!syntax description /Postprocessors/AveragePointSeparation
+
+## Description
+
+This `PostProcessor` computes average distance between a set of points specified in two array `first_point` and `last_point`.
```
[the remark this answers is not in the document]
Comment (jain651) [reply modules/tensor_mechanics/doc/content/source/postprocessors/AveragePointSeparation.md:7]: Done!

File: modules/tensor_mechanics/include/postprocessors/AveragePointSeparation.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "GeneralPostprocessor.h"
+
+/**
+ * Compute the value of a variable at a specified location.
+ *
+ * Warning: This postprocessor may result in undefined behavior if utilized with
```
[the remark this answers is not in the document]
Comment (jain651) [reply modules/tensor_mechanics/include/postprocessors/AveragePointSeparation.h:17]: I have removed it. It should not be here.

File: modules/tensor_mechanics/include/postprocessors/AveragePointSeparation.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+protected:
+  std::vector<VariableName> _displacements;
+
+  /// The vector of displacements number we are operating on
+  std::vector<int> _disp_num;
+
+  /// A reference to the system containing the variable
+  const System & _system;
+
+  /// The point to locate
```
[the remark this answers is not in the document]
Comment (jain651) [reply modules/tensor_mechanics/include/postprocessors/AveragePointSeparation.h:41]: Done!

File: modules/tensor_mechanics/include/postprocessors/AveragePointSeparation.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  const std::vector<Point> & _first_point;
+  const std::vector<Point> & _last_point;
+
+  /// The value of the variable at the desired location
```
[the remark this answers is not in the document]
Comment (jain651) [reply modules/tensor_mechanics/include/postprocessors/AveragePointSeparation.h:45]: Done!

File: modules/tensor_mechanics/src/postprocessors/AveragePointSeparation.C [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  InputParameters params = GeneralPostprocessor::validParams();
+  params.addRequiredParam<std::vector<VariableName>>(
+      "displacements",
+      "The displacements appropriate for the simulation geometry and coordinate system");
+  params.addRequiredParam<std::vector<Point>>("first_point",
+                                              "A list of first points in the numerical domain");
+  params.addRequiredParam<std::vector<Point>>("last_point",
+                                              "A list of last points in the numerical domain");
+  params.addClassDescription(
+      "Compute the average separation between a list of two specified locations");
```
[the remark this answers is not in the document]
Comment (jain651) [reply modules/tensor_mechanics/src/postprocessors/AveragePointSeparation.C:43]: Done!

**Comment by jain651** [comment]
closing as it is more suitable for BlackBear App.

**Pull request closed without merging**